File: backend/app/api/v1/trending.py
"""
热点/趋势API端点
提供市场热点识别和趋势分析功能
"""
from fastapi import APIRouter, Query, HTTPException
from typing import List
from pydantic import BaseModel, ValidationError

from app.services.hotspot_analyzer import hotspot_analyzer
import traceback

# 创建router
router = APIRouter()


# ================================
# Response Models
# ================================


class ScoresBreakdown(BaseModel):
    """热点分数明细"""

    twitter: float
    reddit: float
    price: float
    volume: float
    news: float


class HotspotItem(BaseModel):
    """热点项目"""

    coin_id: str
    symbol: str
    name: str
    market_cap_rank: int
    price_usd: float | None = None
    price_change_24h: float | None = None
    volume_24h: float | None = None
    total_score: float
    scores_breakdown: ScoresBreakdown
    timestamp: str


class HotspotsResponse(BaseModel):
    """热点列表响应"""

    hotspots: List[HotspotItem]
    count: int
    updated_at: str


# ================================
# API Endpoints
# ================================


@router.get(
    "/trending/hotspots",
    response_model=HotspotsResponse,
    summary="获取市场热点",
    description="多维度识别当前最热门的加密货币",
    tags=["Trending"],
    responses={
        200: {
            "description": "成功返回热点列表",
            "content": {
                "application/json": {
                    "example": {
                        "hotspots": [
                            {
                                "coin_id": "bitcoin",
                                "symbol": "BTC",
                                "name": "Bitcoin",
                                "market_cap_rank": 1,
                                "price_usd": 45000.0,
                                "price_change_24h": 5.2,
                                "volume_24h": 30000000000,
                                "total_score": 85.5,
                                "scores_breakdown": {
                                    "twitter": 22.5,
                                    "reddit": 18.0,
                                    "price": 25.5,
                                    "volume": 14.5,
                                    "news": 5.0
                                },
                                "timestamp": "2025-01-26T10:00:00"
                            }
                        ],
                        "count": 1,
                        "updated_at": "2025-01-26T10:00:00"
                    }
                }
            }
        },
        500: {"description": "服务器内部错误"},
    }
)
async def get_hotspots(
    limit: int = Query(default=10, ge=1, le=50, description="返回数量"),
    force_refresh: bool = Query(default=False, description="强制刷新（忽略缓存）"),
) -> HotspotsResponse:
    """
    获取市场热点 - 多维度识别热门加密货币

    该端点使用多个维度的实时数据,自动识别当前市场最热门的加密货币项目。

    **评分算法:**
    热点得分基于5个维度的加权计算（总分100分）：
    - 🐦 Twitter提及量（25%权重）- 社交媒体热度
    - 💬 Reddit讨论量（20%权重）- 社区活跃度
    - 📈 24h价格变化（30%权重）- 市场表现
    - 💰 24h交易量（15%权重）- 流动性指标
    - 📰 新闻数量（10%权重）- 媒体关注度

    **特性:**
    - ⚡ 每15分钟更新一次
    - 📊 多维度综合评分
    - 🔄 支持强制刷新
    - 💾 Redis缓存加速

    **速率限制:**
    - 30次/分钟（基于IP）

    **请求示例:**
    ```bash
    # 获取Top 10热点
    curl "http://localhost:8000/api/v1/trending/hotspots?limit=10"

    # 获取Top 20热点，强制刷新
    curl "http://localhost:8000/api/v1/trending/hotspots?limit=20&force_refresh=true"
    ```

    **响应示例:**
    ```json
    {
      "hotspots": [
        {
          "coin_id": "bitcoin",
          "symbol": "BTC",
          "name": "Bitcoin",
          "market_cap_rank": 1,
          "price_usd": 45000.0,
          "price_change_24h": 5.2,
          "volume_24h": 30000000000,
          "total_score": 85.5,
          "scores_breakdown": {
            "twitter": 22.5,
            "reddit": 18.0,
            "price": 25.5,
            "volume": 14.5,
            "news": 5.0
          },
          "timestamp": "2025-01-26T10:00:00"
        }
      ],
      "count": 1,
      "updated_at": "2025-01-26T10:00:00"
    }
    ```

    **得分解读:**
    - 80-100分: 🔥 极热（强烈关注）
    - 60-79分: 🌡️ 热门（值得关注）
    - 40-59分: 📊 活跃（正常水平）
    - <40分: 📉 冷清（关注度低）

    **用途:**
    - 🎯 发现市场机会
    - 📰 追踪热门话题
    - 💡 投资决策参考
    - 🔔 监控市场动态

    **数据源:**
    - CoinGecko - 价格和交易量
    - Twitter API - 社交媒体数据
    - Reddit API - 社区讨论数据
    - CryptoPanic - 新闻数据
    """
    from datetime import datetime

    try:
        # 获取热点数据
        hotspots = await hotspot_analyzer.get_hotspots(limit=limit, force_refresh=force_refresh)
        
        # 处理空结果
        if not hotspots:
            return HotspotsResponse(
                hotspots=[], count=0, updated_at=datetime.utcnow().isoformat()
            )

        # 转换为响应格式，过滤无效数据
        items = []
        for hotspot in hotspots:
            try:
                # 验证必需字段
                if not hotspot.get("coin_id") or not hotspot.get("symbol") or not hotspot.get("name"):
                    print(f"⚠️ 跳过无效热点项: 缺少必需字段")
                    continue
                
                # 确保scores_breakdown存在
                if "scores_breakdown" not in hotspot:
                    print(f"⚠️ 跳过无效热点项: 缺少scores_breakdown")
                    continue
                
                items.append(HotspotItem(**hotspot))
            except ValidationError as e:
                print(f"⚠️ 跳过无效热点项: {e}")
                continue
            except Exception as e:
                print(f"⚠️ 处理热点项时出错: {e}")
                continue

        return HotspotsResponse(
            hotspots=items, count=len(items), updated_at=datetime.utcnow().isoformat()
        )
    
    except HTTPException:
        # 重新抛出HTTP异常
        raise
    except Exception as e:
        # 记录详细错误信息
        error_msg = str(e)
        print(f"❌ 获取热点数据错误: {error_msg}")
        traceback.print_exc()
        
        # 返回友好的错误响应
        raise HTTPException(
            status_code=500,
            detail=f"热点分析服务暂时不可用，请稍后重试。错误: {error_msg if 'DEBUG' in str(e) else '服务内部错误'}"
        )

File: workers-api/src/routes/trending.ts
/**
 * Trending API Routes
 * Provides trending crypto topics and hotspots
 */

import { Hono } from 'hono'
import type { Env } from '../types/env'
import { getSupabaseClient } from '../lib/supabase'
import { createCoinGeckoClient } from '../lib/coingecko'

const trending = new Hono<{ Bindings: Env }>()

/**
 * GET /hotspots
 * Get trending crypto topics/hotspots
 */
trending.get('/hotspots', async (c) => {
  const limit = parseInt(c.req.query('limit') || '10', 10)
  const forceRefresh = c.req.query('force_refresh') === 'true'

  try {
    // Use service role to bypass RLS for backend queries
    const supabase = getSupabaseClient(c.env, true)

    // Check cache first (if not force refresh)
    if (!forceRefresh && c.env.CACHE) {
      const cacheKey = `trending:hotspots:${limit}`
      const cached = await c.env.CACHE.get(cacheKey)
      if (cached) {
        try {
          return c.json(JSON.parse(cached))
        } catch (parseError) {
          // Cache data is malformed, continue to database query
          console.warn('[Trending] Malformed cache data, falling back to database')
        }
      }
    }

    // Query recent popular topics from messages
    const { data: hotTopics, error } = await supabase
      .from('messages')
      .select('content')
      .eq('role', 'user')
      .order('created_at', { ascending: false })
      .limit(100)

    if (error) {
      console.error('[Trending] Database error:', error)
      return c.json(
        {
          error: {
            code: 'DATABASE_ERROR',
            message: 'Failed to fetch trending data',
            status: 500,
          },
        },
        500
      )
    }

    // Extract crypto keywords and count frequency
    const keywordCounts = new Map<string, number>()
    const cryptoKeywords = [
      'bitcoin', 'btc', 'ethereum', 'eth', 'solana', 'sol',
      'cardano', 'ada', 'polygon', 'matic', 'avalanche', 'avax',
      'polkadot', 'dot', 'chainlink', 'link', 'uniswap', 'uni',
      'bnb', 'ripple', 'xrp', 'dogecoin', 'doge', 'tron', 'trx',
      'near', 'algorand', 'algo', 'cosmos', 'atom'
    ]

    hotTopics?.forEach((msg) => {
      const content = msg.content.toLowerCase()
      cryptoKeywords.forEach((keyword) => {
        if (content.includes(keyword)) {
          keywordCounts.set(keyword, (keywordCounts.get(keyword) || 0) + 1)
        }
      })
    })

    // Sort by frequency and get top keywords
    const topKeywords = Array.from(keywordCounts.entries())
      .sort((a, b) => b[1] - a[1])
      .slice(0, Math.min(limit * 3, 30)) // Fetch more than needed to account for API failures and deduplication

    // Fetch real crypto data from CoinGecko
    const coinGecko = createCoinGeckoClient()
    
    // Track seen coin IDs to avoid duplicates (e.g., 'btc' and 'bitcoin' resolve to same coin)
    const seenCoinIds = new Set<string>()
    
    const hotspotPromises = topKeywords.map(async ([keyword, searchCount]) => {
      try {
        // Resolve keyword to coin ID and fetch market data
        const coinId = (coinGecko as any).resolveCoinId(keyword)
        
        // Skip if we've already seen this coin (handles aliases like 'btc'/'bitcoin')
        if (seenCoinIds.has(coinId)) {
          return null
        }
        seenCoinIds.add(coinId)
        
        const coinData = await coinGecko.getCoinPrice(coinId)

        if ('error' in coinData) {
          console.warn(`[Trending] Failed to fetch data for ${keyword}:`, coinData.message)
          return null
        }

        // Calculate total score based on search count and market rank
        // Higher search count = higher score, lower market cap rank = higher score
        const searchScore = searchCount * 10 // Weight for search frequency
        const rankScore = coinData.market_cap_rank
          ? Math.max(0, 100 - coinData.market_cap_rank) // Top 100 coins get bonus
          : 0
        const totalScore = searchScore + rankScore

        return {
          coin_id: coinId,
          symbol: coinData.symbol,
          name: coinData.name,
          market_cap_rank: coinData.market_cap_rank || 9999,
          price_usd: coinData.price_usd,
          price_change_24h: coinData.price_change_24h,
          volume_24h: 0, // CoinGecko free API doesn't provide volume in this endpoint
          total_score: totalScore,
          scores_breakdown: {
            twitter: 0,
            reddit: 0,
            price: Math.abs(coinData.price_change_24h) * 5, // Price volatility as score
            volume: 0,
            news: searchScore / 2, // Search count as proxy for news activity
          },
          timestamp: new Date().toISOString(),
        }
      } catch (err) {
        console.error(`[Trending] Error processing ${keyword}:`, err)
        return null
      }
    })

    // Wait for all API calls and filter out failures
    const hotspotResults = await Promise.all(hotspotPromises)
    const validHotspots = hotspotResults
      .filter((h): h is NonNullable<typeof h> => h !== null)
      .sort((a, b) => b.total_score - a.total_score) // Sort by total score
      .slice(0, limit) // Get top N

    // If we don't have enough results, fetch default popular coins
    if (validHotspots.length < limit) {
      const defaultCoins = ['bitcoin', 'ethereum', 'solana', 'cardano', 'polygon', 'avalanche-2', 'polkadot', 'chainlink']
      const existingCoinIds = new Set(validHotspots.map(h => h.coin_id))
      
      for (const coinId of defaultCoins) {
        if (validHotspots.length >= limit) break
        if (existingCoinIds.has(coinId)) continue
        
        try {
          const coinData = await coinGecko.getCoinPrice(coinId)
          if ('error' in coinData) continue
          
          validHotspots.push({
            coin_id: coinId,
            symbol: coinData.symbol,
            name: coinData.name,
            market_cap_rank: coinData.market_cap_rank || 9999,
            price_usd: coinData.price_usd,
            price_change_24h: coinData.price_change_24h,
            volume_24h: 0,
            total_score: coinData.market_cap_rank ? Math.max(0, 100 - coinData.market_cap_rank) : 0,
            scores_breakdown: {
              twitter: 0,
              reddit: 0,
              price: Math.abs(coinData.price_change_24h) * 5,
              volume: 0,
              news: 0,
            },
            timestamp: new Date().toISOString(),
          })
          existingCoinIds.add(coinId)
        } catch (err) {
          console.warn(`[Trending] Failed to fetch default coin ${coinId}:`, err)
        }
      }
    }

    const response = {
      hotspots: validHotspots,
      count: validHotspots.length,
      updated_at: new Date().toISOString(),
    }

    // Cache the result for 15 minutes
    if (c.env.CACHE) {
      const cacheKey = `trending:hotspots:${limit}`
      await c.env.CACHE.put(cacheKey, JSON.stringify(response), {
        expirationTtl: 60 * 15, // 15 minutes
      })
    }

    return c.json(response)
  } catch (error) {
    console.error('[Trending] Unexpected error:', error)
    return c.json(
      {
        error: {
          code: 'INTERNAL_ERROR',
          message: 'An unexpected error occurred',
          status: 500,
        },
      },
      500
    )
  }
})

export default trending
